initrd/bin/tpmr

```diff
@@ -0,0 +1,382 @@
+}
+
+# tpm2_seal: Seal a file against PCR values and, optionally, a password.
+# If a password is given, both the PCRs and password are required to unseal the
+# file.  PCRs are provided as a PCR list and data file.  PCR data must be
+# provided - TPM2 allows the TPM to fall back to current PCR values, but it is
+# not required to support this.
+tpm2_seal() {
+	TRACE "Under /bin/tpmr:tpm2_seal"
+	file="$1" #$KEY_FILE
+	handle="$2" # 0x8100000$TPM_INDEX
+	pcrl="$3" #sha256:0,1,2,3,4,5,6,7
+	pcrf="$4"
+	pass="$5"
+	mkdir -p "$SECRET_DIR"
+	bname="`basename $file`"
+
+	DEBUG "tpm2_seal: file=$file handle=$handle pcrl=$pcrl pcrf=$pcrf pass=$([ "$pass" ] && echo "<yes>" || echo "<no>")"
+
+	# Create a policy requiring both PCRs and the object's authentication
+	# value using a trial session.
+	TRIAL_SESSION=/tmp/sealfile_trial.session
+	AUTH_POLICY=/tmp/sealfile_auth.policy
+	rm -f "$TRIAL_SESSION" "$AUTH_POLICY"
+	tpm2 startauthsession -g sha256 -S "$TRIAL_SESSION"
+	# We have to clean up the session
+	trap "cleanup_session '$TRIAL_SESSION'" EXIT
+	# Save the policy hash in case the password policy is not used (we have
+	# to get this from the last step, whichever it is).
+	tpm2 policypcr -Q -l "$pcrl" -f "$pcrf" -S "$TRIAL_SESSION" -L "$AUTH_POLICY"
+	CREATE_PASS_ARGS=()
+	if [ "$pass" ]; then
+		# Add an object authorization policy (the object authorization
+		# will be the password).  Save the digest, this is the resulting
+		# policy.
+		tpm2 policypassword -Q -S "$TRIAL_SESSION" -L "$AUTH_POLICY"
+		# Pass the password to create later.  Pass the sha256sum of the
+		# password to the TPM so the password is not limited to 32 chars
+		# in length.
+		CREATE_PASS_ARGS=(-p "hex:$(echo -n "$pass" | sha256sum | cut -d ' ' -f 1)")
+	fi
+
+	# Create the object with this policy and the auth value.
+	# NOTE: We disable USERWITHAUTH and enable ADMINWITHPOLICY so the
+	# password cannot be used on its own, the PCRs are also required.
+	# (The default is to allow either policy auth _or_ password auth.  In
+	# this case the policy includes the password, and we don't want to allow
+	# the password on its own.)
+	# TODO: Providing the password directly limits it to the size of the
+	# largest hash supported by the TPM (at least 32 chars for sha256)
+	tpm2 create -Q -C "/tmp/$PRIMARY_HANDLE_FILE" \
+		-i "$file" \
+		-u "$SECRET_DIR/$bname.priv" \
+		-r "$SECRET_DIR/$bname.pub" \
+		-L "$AUTH_POLICY" \
+		-S "/tmp/$DEC_SESSION_FILE" \
+		-a "fixedtpm|fixedparent|adminwithpolicy" \
+		"${CREATE_PASS_ARGS[@]}"
+
+	tpm2 load -Q -C "/tmp/$PRIMARY_HANDLE_FILE" -u "$SECRET_DIR/$bname.priv" -r "$SECRET_DIR/$bname.pub" -c "$SECRET_DIR/$bname.seal.ctx"
+	read -s -p "TPM owner password: " key_password
+	echo # new line after password prompt
+	# remove possible data occupying this handle
+	tpm2 evictcontrol -Q -C o -P "$key_password" -c "$handle" 2>/dev/null || true
+	DO_WITH_DEBUG tpm2 evictcontrol -Q -C o -P "$key_password" -c "$SECRET_DIR/$bname.seal.ctx" "$handle"
```

Comment:
DO_WITH_DEBUG outputs $key_password on consoles (Others are DEBUG calls outputting password: <yes>)

Reply:
Yeah good catch, will tweak this, thanks

Reply:
DO_WITH_DEBUG can now mask a specific parameter - traces still show whether it was empty but the value is hidden.  And we don't have to duplicate the whole invocation (which I fear would drift over time).

Arrays are great :100: :grin: 

![Screenshot_20230307_135338](https://user-images.githubusercontent.com/109107914/223528032-13c4397e-3d72-4f5e-bdd1-fce4f8ee704a.png)